cw/cw305/util/spiflash.py

```diff
@@ -4,6 +4,241 @@
+
+
+class SAM3UProgrammer:
+  """Class for programming OpenTitan over the SPI interface of SAM3U on CW305."""
+  def __init__(self, firmware_path):
+    """Inits SAM3UProgrammer with the path of a firmware image."""
+    self._firmware_path = firmware_path
+
+  def run(self, fpga):
+    """Programs OpenTitan over the SPI interface of SAM3U on CW305.
+
+    This implementation has two improvements over the `spiflash` tool in
+    lowRISC/opentitan:
+    * Optimizes the happy path by checking the acknowledgement of frame N-1
+      after transmitting frame N instead of transmitting an extra frame to check
+      each frame's acknowledgement.
+    * Sends an all 0xFF frame as the last frame to avoid cases where the last
+      frame gets corrupted and halts the bootstrapping process.
+
+    Args:
+      fpga: CW305 to be programmed, a ``cw.capture.targets.CW305`` instance.
+    """
+    with _Bootstrap(fpga) as bootstrap:
+      frame_number = 0
+      flash_offset = 0
+      prev_frame = None
+      with open(self._firmware_path, mode='rb') as fw:
+        # Read fixed-size blocks from the firmware image
+        for payload in iter(partial(fw.read, _SpiFlashFrame.PAYLOAD_SIZE), b''):
+          frame = _SpiFlashFrame(frame_number, flash_offset, payload)
+          # We need to make sure that the previous frame is transmitted correctly before
+          # proceeding with the next frame.
+          while True:
+            # Transmit frame N, receive acknowledgement for frame N-1.
+            ack = bootstrap.transfer(frame)
+            if prev_frame and prev_frame.expected_ack != ack:
+              # When OpenTitan encounters a transmission error or an out of
+              # sequence frame, it ignores the current frame sends the ack of
+              # the last successfully received frame. Since transmission errors
+              # can occur in both directions, i.e. the previous frame sent by
+              # SAM3U or the ack sent by OpenTitan, and we can't tell them
+              # apart, we should re-transmit the previous frame and check the
+              # ack sent by OpenTitan.
+              print(f'Error transferring {prev_frame}')
+              ack = bootstrap.transfer(prev_frame)
+              if ack == frame.expected_ack:
+                # Ack sent by OpenTitan was corrupted. Proceed with the next frame.
+                break
+              else:
+                # Previous frame sent by SAM3U was corrupted. Send current frame again.
+                pass
+            else:
+              # Correct acknowledgement received or this is the first frame.
+              break
+          prev_frame = frame
+          frame_number += 1
+          flash_offset += _SpiFlashFrame.PAYLOAD_SIZE
+      # Mark the last frame as expected by OpenTitan to finish bootstrapping.
+      bootstrap.transfer(_SpiFlashFrame(frame_number | (1<<31), flash_offset, b''))
```

Comment:
SG. This just adds an empty frame to end the transfer. Maybe just add additional clarification to the comment above to make sure the reader doesn't miss the details of the operation.

Reply:
Done.